modules/ballerina-native/src/main/java/org/ballerinalang/nativeimpl/actions/data/sql/SQLDatasource.java

```diff
@@ -223,20 +204,94 @@ private void buildDataSource(BMap options) {
+            break;
+        case Constants.DBTypes.POSTGRE:
+            if (port <= 0) {
+                port = Constants.DefaultPort.POSTGRE;
+            }
+            jdbcUrl.append("jdbc:postgresql://").append(hostOrPath).append(":").append(port).append("/").append(dbName);
+            break;
+        case Constants.DBTypes.IBMDB2:
+            if (port <= 0) {
+                port = Constants.DefaultPort.IBMDB2;
+            }
+            jdbcUrl.append("jdbc:db2:").append(hostOrPath).append(":").append(port).append("/").append(dbName);
+            break;
+        case Constants.DBTypes.HSQLDB_SERVER:
+            if (port <= 0) {
+                port = Constants.DefaultPort.HSQLDB_SERVER;
+            }
+            jdbcUrl.append("jdbc:hsqldb:hsql://").append(hostOrPath).append(":").append(port).append("/")
+                    .append(dbName);
+            break;
+        case Constants.DBTypes.HSQLDB_FILE:
+            jdbcUrl.append("jdbc:hsqldb:file:").append(hostOrPath).append(File.separator).append(dbName);
+            break;
+        case Constants.DBTypes.H2_SERVER:
+            if (port <= 0) {
+                port = Constants.DefaultPort.H2_SERVER;
+            }
+            jdbcUrl.append("jdbc:h2:tcp:").append(hostOrPath).append(":").append(port).append("/").append(dbName);
+            break;
+        case Constants.DBTypes.H2_FILE:
+            jdbcUrl.append("jdbc:h2:file:").append(hostOrPath).append(File.separator).append(dbName);
+            break;
+        case Constants.DBTypes.DERBY_SERVER:
+            if (port <= 0) {
+                port = Constants.DefaultPort.DERBY_SERVER;
+            }
+            jdbcUrl.append("jdbc:derby:").append(hostOrPath).append(":").append(port).append("/").append(dbName);
+            break;
+        case Constants.DBTypes.DERBY_FILE:
+            jdbcUrl.append("jdbc:derby:").append(hostOrPath).append(File.separator).append(dbName);
+            break;
+        default:
+            jdbcUrl.append("");
+        }
+        return jdbcUrl.toString();
+    }
+
     private void setDataSourceProperties(BMap options, HikariConfig config) {
         Set<String> keySet = options.keySet();
         for (String key : keySet) {
-            String keyName = key;
-            if (keyName.startsWith(Constants.PoolProperties.DATASOURCE)) {
-                String keyValue = keyName.substring(Constants.PoolProperties.DATASOURCE.length());
-                BValue value = options.get(key);
-                if (value instanceof BString) {
-                    config.addDataSourceProperty(keyValue, value.stringValue());
-                } else if (value instanceof BInteger) {
-                    config.addDataSourceProperty(keyValue, ((BInteger) value).intValue());
-                } else if (value instanceof BBoolean) {
-                    config.addDataSourceProperty(keyValue, Boolean.parseBoolean(value.stringValue()));
-                }
```

Comment:
As per the offline chat, let's fix the int, float type issue in the next release. We will have to check the actual Java data types using reflection for the data source class.